virtual/registry/kv_registry.go

```diff
@@ -250,120 +250,136 @@ func (k *kvRegistry) EnsureActivation(
 				req.ActorID, req.Namespace, errActorDoesNotExist)
 		}
 
-		serverKey := getServerKey(ra.Activation.ServerID)
-		v, ok, err := tr.Get(ctx, serverKey)
-		if err != nil {
-			return nil, err
-		}
-
 		var (
-			server       serverState
-			serverExists bool
+			refs            []types.ActorReference
+			currActivations = ra.Activations
+			isActivated     = make(map[string]bool)
 		)
-		if ok {
-			if err := json.Unmarshal(v, &server); err != nil {
-				return nil, fmt.Errorf("error unmarsaling server state with ID: %s", req.ActorID)
-			}
-			serverExists = true
-		}
+		ra.Activations = []activation{}
 
 		vs, err := tr.GetVersionStamp()
 		if err != nil {
 			return nil, fmt.Errorf("error getting versionstamp: %w", err)
 		}
+		isServerIdBlacklisted := make(map[string]bool)
+		for _, s := range req.BlacklistedServerIDs {
+			isServerIdBlacklisted[s] = true
+		}
 
-		var (
-			currActivation, activationExists = ra.Activation, ra.Activation.ServerID != ""
-			timeSinceLastHeartbeat           = versionSince(vs, server.LastHeartbeatedAt)
-			serverID                         string
-			serverAddress                    string
-			serverVersion                    int64
-		)
-		if activationExists &&
-			serverExists &&
-			timeSinceLastHeartbeat < HeartbeatTTL &&
-			currActivation.ServerID != req.BlacklistedServerID {
-			// We have an existing activation and the server is still alive, so just use that.
-
-			// It is acceptable to look up the ServerVersion from the server discovery key directly,
-			// as long as the activation is still active, it guarantees that the server's version
-			// has not changed since the activation was first created.
-			serverVersion = server.ServerVersion
-			serverID = currActivation.ServerID
-			serverAddress = server.HeartbeatState.Address
-		} else {
-			// We need to create a new activation because either:
-			//   1. There is no activation OR
-			//   2. The server the actor is currently activated on has stopped heartbeating OR
-			//   3. The server the actor is currently activated on has blacklisted this actor (most
-			//       likely for balancing reasons)
-			liveServers, err := getLiveServers(ctx, vs, tr)
+		for _, a := range currActivations {
```

Comment:
Can you comment this section and explain what its doing? Also this entire code block is getting kind of long and wild, any way we can break it into some helpers?